File: Table_menu.m
function Table_menu(hf)

global player


%% Plot Menu
mplot = uimenu(hf,'Text','Plot Settings');
% Create child menu items for the menu 'Selection

uimenu(mplot,'Text','Plot Selected','Checked',player.setting.plot_selected,...
   'Tag','plot_selected', 'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','Plot time diagram','Checked',player.setting.plot_timediagram,...
   'Tag','plot_timediagram', 'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','Extend Shot Line','Checked','off', ...
   'Tag', 'plot_check_extendline', 'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','Plot only in blue','Checked',player.setting.plot_only_blue_table, ...
   'Tag', 'plot_check_bluetable', 'MenuSelectedFcn',@plot_menu_function);


% White Ball
uimenu(mplot,'Separator','on','Text','White Ball','Checked',player.setting.ball(1).ball,...
   'Tag', 'plot_check_white_ball', ...
   'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','White Ball line','Checked',player.setting.ball(1).line,...
   'Tag', 'plot_check_white_line', ...
   'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','White Initial Position','Checked',player.setting.ball(1).initialpos,...
   'Tag', 'plot_check_white_initialpos', 'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','White Ball marker','Checked',player.setting.ball(1).marker,...
   'Tag', 'plot_check_white_marker', ...
   'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','White Ghostball','Checked',player.setting.ball(1).ghostball, ...
   'Tag', 'plot_check_white_ghostball', ...
   'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','White ball Diamond position','Checked',player.setting.ball(1).diamondpos, ...
   'Tag', 'plot_check_white_diamondpos', ...
   'MenuSelectedFcn',@plot_menu_function);


% Yellow Ball
uimenu(mplot,'Separator','on','Text','Yellow Ball','Checked',player.setting.ball(2).ball,...
   'Tag', 'plot_check_yellow_ball', ...
   'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','Yellow Ball line','Checked',player.setting.ball(2).line,...
   'Tag', 'plot_check_yellow_line', ...
   'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','Yellow Initial Position','Checked',player.setting.ball(2).initialpos,...
   'Tag', 'plot_check_yellow_initialpos', 'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','Yellow Ball marker','Checked',player.setting.ball(2).marker,...
   'Tag', 'plot_check_yellow_marker', ...
   'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','Yellow Ghostball','Checked',player.setting.ball(2).ghostball, ...
   'Tag', 'plot_check_yellow_ghostball', ...
   'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','Yellow ball Diamond position','Checked',player.setting.ball(2).diamondpos, ...
   'Tag', 'plot_check_yellow_diamondpos', ...
   'MenuSelectedFcn',@plot_menu_function);

% Red Ball
uimenu(mplot,'Separator','on','Text','Red Ball','Checked',player.setting.ball(3).ball,...
   'Tag', 'plot_check_red_ball', ...
   'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','Red Ball line','Checked',player.setting.ball(3).line,...
   'Tag', 'plot_check_red_line', ...
   'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','Red Initial Position','Checked',player.setting.ball(3).initialpos,...
   'Tag', 'plot_check_red_initialpos', 'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','Red Ball marker','Checked',player.setting.ball(3).marker,...
   'Tag', 'plot_check_red_marker', ...
   'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','Red Ghostball','Checked',player.setting.ball(3).ghostball, ...
   'Tag', 'plot_check_red_ghostball', ...
   'MenuSelectedFcn',@plot_menu_function);
uimenu(mplot,'Text','Red ball Diamond position','Checked',player.setting.ball(3).diamondpos, ...
   'Tag', 'plot_check_red_diamondpos', ...
   'MenuSelectedFcn',@plot_menu_function);


% Other menu
mother = uimenu(hf,'Text','Routes Menu');

% Delete Point
uimenu(mother,'Text','Del Point','MenuSelectedFcn',@ShotEditDelete_menu_function);
% Linethickness
uimenu(mother,'Text','Set route thickness','MenuSelectedFcn',@LineWidth_menu_function);


%% Toolbar
tb = findall(hf,'Type','uitoolbar');
%tb = uitoolbar(hfigure);

% To first
player.pt_First = uipushtool(tb);
player.pt_First.Separator = 'on';
player.pt_First.CData = player.icon.first;
player.pt_First.Tooltip = 'Go to first time step';
player.pt_First.Tag = 'first';
player.pt_First.ClickedCallback = @PlayerFunction;

% big step back
player.pt_FastBackward = uipushtool(tb);
player.pt_FastBackward.CData = player.icon.fastbackward;
player.pt_FastBackward.Tooltip = '1 second back';
player.pt_FastBackward.Tag = 'fastbackward';
player.pt_FastBackward.ClickedCallback = @PlayerFunction;

% 1 step back
player.pt_OneBack = uipushtool(tb);
player.pt_OneBack.CData = player.icon.onebackward;
player.pt_OneBack.Tooltip = '1 step back';
player.pt_OneBack.Tag = 'onebackward';
player.pt_OneBack.ClickedCallback = @PlayerFunction;

% play
player.pt_Play = uitoggletool(tb);
player.pt_Play.CData = player.icon.play;
player.pt_Play.Tooltip = 'Start/Stop play';
player.pt_Play.Tag = 'play';
player.pt_Play.ClickedCallback = @PlayerFunction;

% 1 step forward
player.pt_OneForward = uipushtool(tb);
player.pt_OneForward.CData = player.icon.oneforward;
player.pt_OneForward.Tooltip = '1 step forward';
player.pt_OneForward.Tag = 'oneforward';
player.pt_OneForward.ClickedCallback = @PlayerFunction;

% big step forward
player.pt_FastForward = uipushtool(tb);
player.pt_FastForward.CData = player.icon.fastforward;
player.pt_FastForward.Tooltip = '1 second forward';
player.pt_FastForward.Tag = 'fastforward';
player.pt_FastForward.ClickedCallback = @PlayerFunction;

% to Last
player.pt_Last = uipushtool(tb);
player.pt_Last.CData = player.icon.last;
player.pt_Last.Tooltip = 'Go to end';
player.pt_Last.Tag = 'last';
player.pt_Last.ClickedCallback = @PlayerFunction;

% previous
player.pt_Previous = uipushtool(tb);
player.pt_Previous.Separator = 'on';
player.pt_Previous.CData = player.icon.previous;
player.pt_Previous.Tooltip = 'Go to previous shot';
player.pt_Previous.Tag = 'previous';
player.pt_Previous.ClickedCallback = @PlayerFunction;

% next
player.pt_Next = uipushtool(tb);
player.pt_Next.CData = player.icon.next;
player.pt_Next.Tooltip = 'Go to next shot';
player.pt_Next.Tag = 'next';
player.pt_Next.ClickedCallback = @PlayerFunction;

% record
player.pt_Record = uipushtool(tb);
player.pt_Record.CData = player.icon.record;
player.pt_Record.Tooltip = 'Record to file';
player.pt_Record.Tag = 'record';
player.pt_Record.ClickedCallback = @PlayerFunction;

% CurrentBall
ttImage = zeros(16,16,3);
ttImage(:,:,3) = ones(16);
tt.Icon = ttImage;
player.pt_CurrentBall = uipushtool(tb);
player.pt_CurrentBall.Separator = 'on';
player.pt_CurrentBall.CData = player.icon.WhiteBall;
player.pt_CurrentBall.Tooltip = 'Color of current cue ball';
player.pt_CurrentBall.Tag = 'CurrentBall';

% DrawWhiteLines
player.pt_DrawWhiteLines = uipushtool(tb);
player.pt_DrawWhiteLines.CData = player.icon.drawwhitelines;
player.pt_DrawWhiteLines.Tooltip = 'Draw annotions line for white ball';
player.pt_DrawWhiteLines.Tag = 'drawwhitelines';
player.pt_DrawWhiteLines.ClickedCallback = @PlayerFunction;

% DrawYellowLines
player.pt_DrawYellowLines = uipushtool(tb);
player.pt_DrawYellowLines.CData = player.icon.drawyellowlines;
player.pt_DrawYellowLines.Tooltip = 'Draw annotions line for white ball';
player.pt_DrawYellowLines.Tag = 'drawyellowlines';
player.pt_DrawYellowLines.ClickedCallback = @PlayerFunction;

% DrawRedLine
player.pt_DrawRedLines = uipushtool(tb);
player.pt_DrawRedLines.CData = player.icon.drawredlines;
player.pt_DrawRedLines.Tooltip = 'Draw annotions line for white ball';
player.pt_DrawRedLines.Tag = 'drawredlines';
player.pt_DrawRedLines.ClickedCallback = @PlayerFunction;

File: table_menu.py
from PyQt5.QtWidgets import QMenu, QAction

def create_table_menu(main_window, player):
    """
    Creates the table menu for the main window.

    Args:
        main_window: The main application window.
        player: The player object containing settings and state.
    """
    menu_bar = main_window.menuBar()

    # Plot Menu
    plot_menu = QMenu("Plot Settings", main_window)
    menu_bar.addMenu(plot_menu)

    plot_selected_action = QAction("Plot Selected", main_window)
    plot_selected_action.setCheckable(True)
    plot_selected_action.setChecked(player['setting']['plot_selected'])
    plot_selected_action.triggered.connect(lambda: toggle_setting(player, 'plot_selected'))
    plot_menu.addAction(plot_selected_action)

    plot_time_diagram_action = QAction("Plot Time Diagram", main_window)
    plot_time_diagram_action.setCheckable(True)
    plot_time_diagram_action.setChecked(player['setting']['plot_timediagram'])
    plot_time_diagram_action.triggered.connect(lambda: toggle_setting(player, 'plot_timediagram'))
    plot_menu.addAction(plot_time_diagram_action)

    extend_shot_line_action = QAction("Extend Shot Line", main_window)
    extend_shot_line_action.setCheckable(True)
    extend_shot_line_action.setChecked(False)  # Default value
    extend_shot_line_action.triggered.connect(lambda: toggle_setting(player, 'plot_check_extendline'))
    plot_menu.addAction(extend_shot_line_action)

    plot_only_blue_action = QAction("Plot Only in Blue", main_window)
    plot_only_blue_action.setCheckable(True)
    plot_only_blue_action.setChecked(player['setting']['plot_only_blue_table'] == 'on')
    plot_only_blue_action.triggered.connect(lambda: toggle_setting(player, 'plot_only_blue_table', toggle_value='on'))
    plot_menu.addAction(plot_only_blue_action)

    # Ball Settings
    for color, index in zip(['White', 'Yellow', 'Red'], range(3)):
        ball_menu = QMenu(f"{color} Ball", main_window)
        plot_menu.addMenu(ball_menu)

        ball_action = QAction(f"{color} Ball", main_window)
        ball_action.setCheckable(True)
        ball_action.setChecked(player['setting']['ball'][index]['ball'])
        ball_action.triggered.connect(lambda checked, idx=index: toggle_ball_setting(player, idx, 'ball', checked))
        ball_menu.addAction(ball_action)

        line_action = QAction(f"{color} Ball Line", main_window)
        line_action.setCheckable(True)
        line_action.setChecked(player['setting']['ball'][index]['line'])
        line_action.triggered.connect(lambda checked, idx=index: toggle_ball_setting(player, idx, 'line', checked))
        ball_menu.addAction(line_action)

        initial_pos_action = QAction(f"{color} Initial Position", main_window)
        initial_pos_action.setCheckable(True)
        initial_pos_action.setChecked(player['setting']['ball'][index]['initialpos'])
        initial_pos_action.triggered.connect(lambda checked, idx=index: toggle_ball_setting(player, idx, 'initialpos', checked))
        ball_menu.addAction(initial_pos_action)

        marker_action = QAction(f"{color} Ball Marker", main_window)
        marker_action.setCheckable(True)
        marker_action.setChecked(player['setting']['ball'][index]['marker'])
        marker_action.triggered.connect(lambda checked, idx=index: toggle_ball_setting(player, idx, 'marker', checked))
        ball_menu.addAction(marker_action)

        ghostball_action = QAction(f"{color} Ghostball", main_window)
        ghostball_action.setCheckable(True)
        ghostball_action.setChecked(player['setting']['ball'][index]['ghostball'])
        ghostball_action.triggered.connect(lambda checked, idx=index: toggle_ball_setting(player, idx, 'ghostball', checked))
        ball_menu.addAction(ghostball_action)

        diamond_pos_action = QAction(f"{color} Ball Diamond Position", main_window)
        diamond_pos_action.setCheckable(True)
        diamond_pos_action.setChecked(player['setting']['ball'][index]['diamondpos'])
        diamond_pos_action.triggered.connect(lambda checked, idx=index: toggle_ball_setting(player, idx, 'diamondpos', checked))
        ball_menu.addAction(diamond_pos_action)

def toggle_setting(player, setting_key, toggle_value=True):
    """
    Toggles a boolean setting in the player object.

    Args:
        player: The player object.
        setting_key: The key of the setting to toggle.
        toggle_value: The value to toggle to (default is True).
    """
    current_value = player['setting'].get(setting_key, False)
    player['setting'][setting_key] = not current_value if toggle_value is True else toggle_value
    print(f"Toggled {setting_key} to {player['setting'][setting_key]}")

def toggle_ball_setting(player, ball_index, setting_key, value):
    """
    Toggles a setting for a specific ball in the player object.

    Args:
        player: The player object.
        ball_index: The index of the ball (0, 1, or 2).
        setting_key: The key of the setting to toggle.
        value: The value to set.
    """
    player['setting']['ball'][ball_index][setting_key] = value
    print(f"Set {setting_key} for ball {ball_index} to {value}")